File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
/**
 * 0: 벽이 있는 자리, 1: 벽이 없는 자리
 * 시작 지점: (1, 1)
 */
function solution(maps) {
    const n = maps.length
    const m = maps[0].length
    const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]

    const bfs = (start) => {
        const queue = [start]
        const visited = Array.from({ length: n }, () => Array(m).fill(false));
        visited[0][0] = true;

        while (!!queue.length) {
            const [y, x, dist] = queue.shift()
            if(y === n - 1 && x === m - 1) return dist
            
            for(const [dy, dx] of directions) {
                const nx = x + dx
                const ny = y + dy
                
                // 맵 넘어가면 안감
                const isOver = nx < 0 || nx >= m || ny < 0 || ny >= n
                
                if (!isOver) {
                    // 벽이면 안감
                    const isWall = maps[ny][nx] === 0
                    // 방문했으면 안감
                    const isVisited = visited[ny][nx]
                    
                    if(!isWall && !isVisited) {
                        visited[ny][nx] = true
                        queue.push([ny, nx, dist+1])
                    }
                }

            }
        }
        
        return -1
    }

    return bfs([0, 0, 1])
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque


    
def solution(maps):
    n = len(maps)    # 행 (y)
    m = len(maps[0]) # 열 (x)
    
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]

    visited = [[False for _ in range(m)] for _ in range(n)]
    
    # BFS 시작
    queue = deque([(0, 0, 1)])
    visited[0][0] = True
    
    # 큐가 끝났는데도 도착 못하면 -1 return
    while queue:
        # 큐에서 한개 꺼내기
        y, x, dist = queue.popleft()
        
        
        # 만약 꺼낸 곳이 도착지라면 return
        if y == n - 1 and x == m - 1:
            return dist
        
        # 현재 위치에서 갈 수 있는 곳 다 가보기
        for i in range(4):
            nx = x + dx[i] 
            ny = y + dy[i]

            # 맵 범위 내에 있고
            if 0 <= ny < n and 0 <= nx < m:
                # 벽이 아니며(1), 아직 방문하지 않았다면
                if maps[ny][nx] == 1 and not visited[ny][nx]:
                    visited[ny][nx] = True
                    # 다음 칸으로 이동할 때 거리를 +1 해서 큐에 삽입
                    queue.append((ny, nx, dist + 1))
    
    return -1
        
        